### PR TITLE
GetSource will go to the host and can reentrant to engine.

### PR DESCRIPTION
**Pull request by akroshg**
Our exception handling did not like that. Fixed that by putting macro to satisfy that.

### CODE REVIEW

**Comment by akroshg** [comment]
I have made quick and dirty changes in the jshost to verify the assert is fixed. Although there is not a easy way to add a test for this.

**Comment by akroshg** [comment]
@curtisman @digitalinfinity please have a look at this

**Comment by digitalinfinity** [comment]
Looks fine to me- this looks very similar to a bug that @irinayat-MS fixed last week- we should probably try to consolidate those two codepaths

**Comment by akroshg** [comment]
@irinayat-MS please have a look at this.

File: lib/Runtime/Base/Utf8SourceInfo.cpp [hunk 1]
```diff
@@ -59,6 +59,25 @@ namespace Js
 #endif
     }
 
+    LPCUTF8 Utf8SourceInfo::GetSourceFromHolder(const char16 * reason) const
+    {
+        LPCUTF8 toReturn = nullptr;
+        if (GetScriptContext()->GetThreadContext()->IsScriptActive())
+        {
+            BEGIN_LEAVE_SCRIPT(GetScriptContext())
```
Comment (irinayat-MS) [inline lib/Runtime/Base/Utf8SourceInfo.cpp]: The fix I had in mind was to add these guards in DynamicSourceHolder when it makes external calls (the hits we are getting from the crawler). Are you seeing re-entrancy with JsrtSourceHolder?

**Comment by akroshg** [comment]
@irinayat-MS it is from crawler only. but there is no access to ScriptContext from DynamicSourceHolder, that's why I have done here.

**Comment by irinayat-MS** [comment]
As far as I understand, adding guards to DynamicSourceHolder gives tighter bounds for when we exit/enter script, than adding them here. For my fix I had to cache ScriptContext on the dynamic holder. If other source holders don't call external code, it seems a better fix to me. What do you think?

---
In reply to: [395175975](https://github.com/Microsoft/ChakraCore/pull/5260#issuecomment-395175975) [](ancestors = 395175975)

**Comment by akroshg** [comment]
fine with me. Is it guaranteed that JsrtSourceHolder cannot be extended and re-entrant?

**Comment by irinayat-MS** [comment]
no such guarantee, but isn't it responsibility of the code that *makes* the external call to guard it? its clients shouldn't know about the implementation details, should they?

---
In reply to: [395182510](https://github.com/Microsoft/ChakraCore/pull/5260#issuecomment-395182510) [](ancestors = 395182510)

**Comment by akroshg** [comment]
Closing this to favor another fix.

**Pull request closed without merging**